File: 09/09.py
import heapq
import math
import sys

import kstricks.grid as grid


def main():
    g = grid.Grid.from_rows(map(int, line.strip()) for line in sys.stdin)

    low = {p: v for p, v in g.items() if all(g[q] > v for q in g.neighbors(p))}
    print(sum(low.values()) + len(low))

    basins = {p: [p] for p in low}
    seen = {p: p for p in low}
    frontier = [*low.keys()]
    while frontier:
        p = frontier.pop()
        for q in g.neighbors(p):
            if q in seen or g[q] == 9:
                continue
            frontier.append(q)
            seen[q] = seen[p]
            basins[seen[p]].append(q)
    print(math.prod(heapq.nlargest(3, map(len, basins.values()))))


if __name__ == "__main__":
    main()

File: 09/09.jl
using DataStructures

grid = [[parse(Int, c) for c in row] for row in eachline()]

Δ = [(0, 1), (0, -1), (1, 0), (-1, 0)]

islow(i, j) = all(Δ) do δ
  i′, j′ = (i, j) .+ δ
  (
    !(1 ≤ i′ ≤ length(grid) && 1 ≤ j′ ≤ length(grid[i]))
    || grid[i][j] < grid[i′][j′]
  )
end

p1() = sum(
  1+c
  for (i, row) in enumerate(grid)
  for (j, c) in enumerate(row)
  if islow(i, j)
)

p2() = begin
  roots = [(i, j) for (i, row) in enumerate(grid) for (j, _) in enumerate(row) if islow(i, j)]
  sizes = Dict(p => 1 for p in roots)
  basins = Dict(p => p for p in roots)

  frontier = collect(roots)
  while !isempty(frontier)
    parent = pop!(frontier)
    for δ in Δ
      i′, j′ = child = parent .+ δ
      if (
          !(1 ≤ i′ ≤ length(grid) && 1 ≤ j′ ≤ length(grid[i′]))
          || grid[i′][j′] == 9
          || child ∈ keys(basins)
      )
        continue
      end
      root = basins[child] = basins[parent]
      sizes[root] += 1
      push!(frontier, child)
    end
  end

  prod(partialsort!(collect(values(sizes)), 1:3, rev=true))
end

println((p1(), p2()))
